### ISSUE TITLE
assigning loads processes in intervals

### ISSUE THREAD

**Issue by josep-m-carbonell** [KratosMultiphysics/Kratos#216]
If I apply a load in an interval what must this process do when the interval ends, keep the load or unassign the load?

Is there any field to select the wished behavior in a assignment process?

@KratosMultiphysics/technical-committee

**Comment by pooyan-dadvand** [comment]
In my opinion it should remove the load but I understand that it may lead to very rare overal behaviour. For example if we apply the following loads to the same node:

- Load= 10 in t=[1,2]
- Load = 20 in t=[1.9, 3]

if we remove the load after 2s due to the first interval it will interrupt the second load. This can be avoided if the process applies the load in each solution step but I don't like this work around

**Comment by maceligueta** [comment]
Is it possible to create the loads at the beginning of the time step and remove them all at the end of the time step? It would be a robust method, in my opinion. @RiccardoRossi proposed this approach for the boundary conditions in general. If you choose this way, you never have to unassign loads selectively (which needs detection of the end of the interval). It's a 'brute force' approach, but quite robust.
[Sorry @pooyan-dadvand , now I see that your last sentence refers to this, but I still like the idea]

**Comment by josep-m-carbonell** [comment]
@maceligueta is the easier way to solve a problem, but not the better way in my opinion. The skill worth more than brute force. Furthermore, in this case, the cost of assigning and unassigning is not so negligible as fixing and unfixing a dof.

**Comment by RiccardoRossi** [comment]
In my opinion for consistency with the fixing behaviour we should reset the load after the interval.
I think that the suggestion of MA makes complete sense on this. Also applying the loading at the beginning every time at the beginning of the interval will work if the mesh is changed, or if the domain is rebalanced in mpi, while alternatives would not. 

concerning performance, the unloading is very cheap, since it implies just calling a "clear" function on the load vector on each condition. 
What is potentially more expensive is the evaluation of space-dependent loads.
i think there are the following typical scenarios:

1 - applying a constant load. Here performance should not be a problem, and the process is completely parallel
2 - applying a loading which depends on time but not on the space. Also here performance should be no issue
3 - applying a time dependent loading which depends on space. The evaluation of the function is not parallel, anyway we do need to recalculate the load at each step so no real problem
4 - applying a space-dependent load with no time dependency - this is the only case in which applying the load is significatively more expensive. Note that this can be done mpi-parallel but not openmp parallel due to python threading  limitations. 

I still believe that even for the case 4 the cost should be small when compared to the actual system solution. If you like i can try this in the "experimental_processes" branch so that we can evaluate it

**Comment by josep-m-carbonell** [comment]
For loading consistency, the value of the load must be applied to condition data_value  via SetValue. I don't know if it is so cheap. Then, there is a direct relationship between the load value and the condition. That  means two different loads has independent conditions over the same geometry. Clearing one of them must not affect to   the other. These means that the only behavior that you would like to define is clearing the load after the interval end or leaving it assigned with its last value. 

The other thing is, consider the case where you are sure that nothing is going to change along time during the proceses execution. In that case I would like to deactivate the 'brute force' of unassigning  everything at each step. A system is needed to have this capability.

**Comment by pooyan-dadvand** [comment]
Can we have an apply load process which sorts the loads by their end time and apply the load or remove it at initial solution step?

**Comment by RiccardoRossi** [comment]
[edit: correcting wrongly typed text]

What if the mesh changed in one step? To handle that case you need to reapply the loads at the beginning of the Next step.

Also i now think It is not a good idea to wipe the loads at the end of the step (i changed my mind on this). My reasoning is that if we do the same with nodal values we would "change the past" for the following time  steps. That is, we would  not be able in the following steps to compute correctly things that depende on the history, say acceleration.

**Comment by RiccardoRossi** [comment]
on the top of my other comment, 
       another problem is that when you are at the end of a step, you do not know if the dt will be changed in the following, so you are not able to know if the following time step is out of the interval. this implies that you do not know if you should do something or not.

**Comment by josep-m-carbonell** [comment]
This about meshing is not a problem you don't need to reapply the loads. If you insert new entities you must clone data.  How are you changing the meshes?

**Comment by josep-m-carbonell** [comment]
If the step is changed, it is not a problem, at the beginning the load will be reassigned again, because you are still inside of the interval.

**Comment by RiccardoRossi** [comment]
If you use an external mesher, It is definitely needed to reapply the BCs.

For example the only thing respected by mmg mesher (which freely modifies
the boundary respectivo the shape) is the identificación of the boundary
faces, which is enough to reconstruct the submodelparts they belong to

Advantage is that once you are guaranteed that the modelparts are preserved
you can reapply BCs.

**Comment by josep-m-carbonell** [comment]
I do not agree with you @RiccardoRossi, if you call a mesher, then you have to rebuild things correctly. It is **not** "definitely needed" to reassign the BC to all domain. You transfer the data from the old entities to the new entities (this data contains assigned values for the boundary conditions). The transfer is independent of assigning boundary conditions. I think that you are mixing different operations.

**Comment by RiccardoRossi** [comment]
@josep-m-carbonell , then let me rephrase. It is **highly convenient** (although may not be mandatory) to have BCs reapplied on submodelparts in the case in which the mesh is changed.

the thing is that generally meshers allow one to specify flags to be mantained on the faces of the domain (or even on portions of the volume). This is surely the case for triangle, tetgen and mmg. It is relatively easy to make such flags to map identically to the submodelparts which are present in the original model, so that conditions can be applied very naturally on the remeshed domain even in the case in which the boundaries are modified. If this is not done, one would have to somehow interpolate the boundaries of the new mesh to the old one. This may well be problematic in the case in which the functions to be applied are not continunous, or if the mesh is curve, since new added nodes may be positioned so to lay on a reconstructed curved mesh instead of laying exactly on the faces on the previous one. This for example is a quite powerful feature provided by the MMG lib in the 3D case, but one will obviously encounter the same issue if refining a quadratic mesh. 

An even more problematic feature appears if someone tries to apply a nodal condition via refining an existing mesh. Consider for example the case in which a tet has all the nodes on the boundary. if the tet is refined, one would inevitably apply the same value also on the newly added node. However, at least one of the edges may not belong to the volume, so that applying the boundary value to the node that appears from the refinement of such edge is simply wrong. 
Now detecting such case is very difficult if you have nodal data alone, however the case is handled in a completely natural way if you use the boundary faces in reapplying the BCs.

my point is that all of such issues are avoided by simply reapplying the BCs at the beginning of the step, so i am strongly in favor of doing that systematically

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi I think that the meshing BC problem is a next step consideration. Now I would focus more on how we avoid conflicts in apply BC via intervals to the same mesh. 

Just to resume we have 3 possibilities until now:
1. Checking the processes for conflicts at the beginning and stop if two overlapping intervals act on same SubModelPart

2. Apply and remove them in each time step
3. Reorder them by the end interval time.

IMO we can at least start with the first one for now. (may be even from GUI)

**Comment by RiccardoRossi** [comment]
I would add 4, which would be current my vote (as i told you this is for consistencia with the behaviour of nodal values):

4 - apply loads at the beginning of the interval and leave them applied outside of It (if you want them to go to zero you can simply assign by table that they go to zero before the end of the interval)

**Comment by josep-m-carbonell** [comment]
@RiccardoRossi , about the option 4) you have exposed : I don't think this matches with the concept of a load applied in an interval. If you press something during 3 seconds, means that you release the pressure once the 3 seconds have past. It is not consistent with the behavior of other boundary conditions, but maybe it does not have to be like this.

**Comment by RiccardoRossi** [comment]
let's imagine we want to have  the following loading case:

 - time 0.0 to 1.0  variable LOAD set to "(x+y)*t" on all the conditions
 - time 1.0 to 2.0 variable LOAD set to zero
 - time 2.0 to 3.0 variable LOAD set to "5.0"

in my proposal this would correspond to writing the following in the ProjectParameters:

        {
                "process_list" : [
                    {
                        "python_module"   : "assign_scalar_field_to_conditions_process",
                        "kratos_module" : "KratosMultiphysics",
                        "process_name"          : "AssignScalarFieldToConditionsProcess",
                        "Parameters"            : {
                            "model_part_name":"Main",
                            "variable_name": "LOAD",
                            "value" : "(x+y)*t",
                            "interval" : [0.0, 1.0]
                        }
                    },
                    {
                        "python_module"   : "assign_scalar_field_to_conditions_process",
                        "kratos_module" : "KratosMultiphysics",
                        "process_name"          : "AssignScalarFieldToConditionsProcess",
                        "Parameters"            : {
                            "model_part_name":"Main",
                            "variable_name": "LOAD",
                            "value" : "0.0",
                            **"apply_only_once" : true**
                            "interval" : [1.0, 2.0]
                        }
                    },
                    {
                        "python_module"   : "assign_scalar_field_to_conditions_process",
                        "kratos_module" : "KratosMultiphysics",
                        "process_name"          : "AssignScalarFieldToConditionsProcess",
                        "Parameters"            : {
                            "model_part_name":"Main",
                            "variable_name": "LOAD",
                            "value" : "5.0",
                            "interval" : [2.0, 3.0],
                            **"apply_only_once" : true**
                        }
                    }
                ]
            }

note that in this case the only variable that can be applied only once is the 5.0. my suggestion is to introduce an OPTIONAL setting **apply_only_once** which defaults to false. This would allow to optimize that specific use case.

Can u guys describe the input for the format you propose so to have a base on the top of which we can discuss?

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand i disagree about your "next step" comment. Thing is that i believe that the base processes should be as generic as possible, so they should definitely take into account the possibility that the mesh can change from one step to the next. Note that remeshing is not the only use case in which this happens. One may for example simply add or remove elements from the loading modelparts, without having to effectively change the mesh. These cases can only be handled if the loading is reapplied at every time step.
A use case for this is an extrusion process in which the mesh does not change but the portion in contact with the air evolves.
